File: code/ch07/7.4.4_single_number.py
class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        counts = [0] * 32
        # 统计 nums 中每个元素的二进制中，对应位上的 1 的个数
        for num in nums:
            for j in range(32):
                counts[j] += num & 1
                num >>= 1
        res, m = 0, 3
        # 对每一位上，统计的 1 的个数取余
        for i in range(32):
            res <<= 1
            res |= counts[31 - i] % m
            
        return res if counts[31] % m == 0 else ~(res ^ 0xffffffff)

File: code/ch07/7.4.4_single_number.cpp
class Solution {
public:
    int singleNumber(vector<int>& nums) {
        vector<int> t(32);
        
        for (int i = 0; i < nums.size(); i++) {
            int num = nums[i];
            for (int j = 31; j >= 0; j--) {
                t[j] += num & 1;
                num >>= 1;
                if (!num) break;
            }
        }
        
        int res = 0;
        for (int j = 31; j >= 0; j--) {
            int num = t[j] % 3;
            if (num) res += 1 << (31 - j);
        }
        
        return res;
    }
};
